### ISSUE TITLE
Feature request: Support for DJI Mini3 DNGs lens correction

### ISSUE THREAD

**Issue by WiLars** [darktable-org/darktable#13319]
**Is your feature request related to a problem? Please describe.**
The DJI Mini3 drone generates DNGs that have embedded lens correction profiles. These profiles are not applied on import or in the lens correction module. Correction using Lensfun is not possible, because the vignette is not only related to luminance but also has a significant color shift (/lensfun/lensfun/issues/1963). This color shift basically makes the raws unusable in Darktable.

**Describe the solution you'd like**
Apply the lens profiles embedded in the DNG files in the context of the lens correction module

**Alternatives**
Preprocessing the DNG files, so that Darktable doesn't have to apply correction profiles. I'm actually not sure how I could do that.

**Additional context**
I think the foundation for this was implemented in darktable-org/darktable/pull/12714 . This functionality works great for my camera, but not for DJI Mini3. 
@paolodepetrillo, do you have some additional insights why this does not work? As you shortly indicated in [this comment](https://github.com/darktable-org/darktable/pull/12714#issuecomment-129929954) you already looked into DJI drones? It looks like you have a deep understanding of OpcodeList in the DNG file format.

**Comment by Nilvus** [comment]
@sgotti: what's your view on that?

@WiLars: given your FR, darktable version related actually seems obvious but think about precising it anyway in future FR/issues. There's a template that request this and it's there on purpose. Thanks!

**Comment by WiLars** [comment]
@Nilvus thanks for the hint. I think in the template for feature requests doesn't request that info, so i wasn't aware.
Anyway, here is some additional info, as found in the bugreport template:

**Platform**
* darktable version : 4.2.0
* OS                : Linux Kernel 5.15.85-1-MANJARO
* Linux - Distro    :Manjaro
* Memory            : 32 Gb
* Graphics card     : Nvidia RTX 3060Ti
* Graphics driver   : 525.60.11
* OpenCL installed  : no
* OpenCL activated  : no
* Xorg              : X.Org X Server 1.21.1.6
* Desktop           : KDE
installed from repo, so i think built parameters do not really apply?


**Additional context**
 - Can you reproduce with a RAW or Jpeg or both? **only raw, in jpeg lens corrections are already applied by drone**
 - Are the steps above reproducible with a fresh edit (i.e. after discarding history)? **yes**
 - If the issue is with the output image, attach an XMP file if (you'll have to change the extension to `.txt`)
 - Is the issue still present using an empty/new config-dir (e.g. start darktable with --configdir "/tmp")? **yes**
 - Do you use lua scripts? **no**

I think i can't just upload a raw file as example here, but if anyone is interested I can provide some.

**Comment by Nilvus** [comment]
> @Nilvus thanks for the hint. I think in the template for feature requests doesn't request that info, so i wasn't aware. Anyway, here is some additional info, as found in the bugreport template:

You're right because for feature request, most of infos of bugreport template are not needed (included for most FR requests) darktable version. Your request is something between camera support and feature request. And for camera support, we often have some request with existing issues or fixed in a previous release.

**Comment by kmilos** [comment]
Please upload samples to raw.pixls.us, and also share here via some cloud service.

**Comment by kmilos** [comment]
So, there is
```
OpcodeList3                     : FixVignetteRadial, GainMap
```

**Comment by spaceChRiS** [comment]
Related: https://github.com/darktable-org/darktable/pull/12880
As GainMap is already implemented for OpcodeList 2, but is here used in OpcodeList 3, the OpcodeList would have to be distinguished in the metadata if I understand it correctly.

**Comment by kmilos** [comment]
> As GainMap is already implemented for OpcodeList 2, but is here used in OpcodeList 3, the OpcodeList would have to be distinguished in the metadata if I understand it correctly.

Not only - it has to be applied in the correct place in the pipeline - OpcoleList2 comes after and is applied to single-channel raw CFA data (currently in the rawprepare module right after the file is opened); OpcodeList3 must be applied just after demosaic and applied to 3 color channels, and this is currently not implemented.

We can't even apply the FixVignetteRadial OpcodeList3 in the lens correction module, as it is in the incorrect place in the pipeline...

**Comment by jenshannoschwalm** [comment]
I have pending work related to dng lens correction and indeed **that** should be as early as possible after demosaicing (or sort of Post-Processing there)

**Comment by kmilos** [comment]
It gets somewhat more complicated - [this sample](https://www.dpreview.com/sample-galleries/2787821900/dji-mini-3-sample-gallery/5855728345) has a (weird) 2x digital zoom: crops out middle 2000x1500, then scales pixels by 2x to get back to "expected" 4000x3000 output...

What is not clear is when the OpcodeList3 is applied - it could be to the whole active area before this crop+rescale because the FixVignetteRadial is unchanged? (The GainMap is again on a 4000x3000 size, but that doesn't tell me it's before or after.)

**Comment by paolodepetrillo** [comment]
@kmilos From that sample gallery of 4 images, some have the 2x digital zoom but [1xsingle.DNG](https://www.dpreview.com/sample-galleries/2787821900/dji-mini-3-sample-gallery/1464899771) does not. The GainMaps for all of them regardless of the 1x / 2x zoom are similar, which indicates that the OpcodeList3 should be applied to the entire image before the crop. If it was intended to be applied after the crop, the GainMap for the 2x image would look like a cropped version of the GainMap from the 1x image.

**Comment by kmilos** [comment]
Indeed, as the identical FixVignetteRadial suggests as well...

**Comment by paolodepetrillo** [comment]
@WiLars As a short term workaround until this feature is implemented in darktable, you can use my dngpreprocess utility to apply the corrections outside of darktable, resulting in a new dng file that already has the corrections applied and can be edited in darktable.

Instructions to build and use are here: https://discuss.pixls.us/t/pixel-7-pro-raw-processing/33942/32

The "-3" option must be specified to apply the OpcodeList3 processing for the DJI files:
`dngpreprocess -3 DJI_9999.dng`

**Comment by WiLars** [comment]
@paolodepetrillo thank you for the hint, just tried it, and with your dngpreprocess the raws look much better!

I addded some exported files (jpeg) to the [folder](https://drive.google.com/drive/folders/1NuFsGJccJ8hERduykU0p8snE1yUH38gK). There is still some shade on top and bottom and some slight color shift, but I suppose this is related to the quality of the embedded profiles...

**Comment by andersnygaard** [comment]
DJI just released it's Air 3 drone using the same 24mm camera and sensor as the Mini 3. I get close to if not the same result from importing the pictures from this camera. However, the drone has a secondary 70mm camera built on the same CMOS architecture and pictures from this camera does work in Darktable, applying lens correction just fine.

Compare files from the two uploaded to my Onedrive at https://1drv.ms/f/s!AtaerbceABlrie5qp45Iwrg9JNABRQ?e=EPWHsF

**Comment by jchnkl** [comment]
> @WiLars As a short term workaround until this feature is implemented in darktable, you can use my dngpreprocess utility to apply the corrections outside of darktable, resulting in a new dng file that already has the corrections applied and can be edited in darktable.
> 
> Instructions to build and use are here: https://discuss.pixls.us/t/pixel-7-pro-raw-processing/33942/32
> 
> The "-3" option must be specified to apply the OpcodeList3 processing for the DJI files: `dngpreprocess -3 DJI_9999.dng`

Is there any ongoing work to incorporate this in darktable / libraw? With the latest darktable version I have access to (4.4.2) there is a significant difference between the images before and after running them through @paolodepetrillo 's `dngpreprocess`:

<details>
  <summary>Screenshot</summary>

![image](https://github.com/darktable-org/darktable/assets/1392715/9c769048-18ab-4138-bacf-a54b1f289ff2)
</details>

Unfortunately this roughly doubles the file size and requires me to store a second file together with the original which is not an ideal situation.

**Comment by jchnkl** [comment]
> > Is there any ongoing work to incorporate this in darktable / libraw?
> 
> Not as far as i know. Bye me a new leica using the opcode3 on it's dng files and i will implement it immediately :-)

A DJI Mini 3 will do ;)

Joking aside: Could you provide me with some starting points where the relevant code is located (libraw or darktable, maybe even on file level?). Then I could check if it's feasible to adapt the code from the `dngpreprocess` util.

Or would that be a no go anyway because it links to the Adobe SDK?

**Comment by paolodepetrillo** [comment]
I don't think it would be worthwhile to refer to the Adobe DNG SDK code from dngpreprocess - the GainMap is a simple operation that's fully documented in the [Adobe DNG Spec](https://helpx.adobe.com/content/dam/help/en/photoshop/pdf/DNG_Spec_1_7_0_0.pdf).

The first step is to read the OpcodeList3 GainMaps out of the file and store them in `dt_image_t` to make them available to processing modules. There is an existing field `dng_gain_maps` that holds the OpcodeList2 GainMaps and a new field would be added for the OpcodeList3 ones. In [dng_opcode.c](https://github.com/darktable-org/darktable/blob/master/src/common/dng_opcode.c), `dt_dng_opcode_process_opcode_list_3` would also process GainMaps like `dt_dng_opcode_process_opcode_list_2` does.

Then we need to figure out where in the processing pipeline to apply the correction. It could be done inside [lens correction](https://github.com/darktable-org/darktable/blob/master/src/iop/lens.cc) as it's at the correct place in the pipeline, although that module is already complicated.. or it could be done in a new module.

Another possibility is to convert the post-demosaic OpcodeList3 GainMap to a pre-demosaic OpcodeList2 GainMap having almost the exact same effect, and apply it using the existing implementation in [rawprepare](https://github.com/darktable-org/darktable/blob/master/src/iop/rawprepare.c). This might not be strictly compliant with the spec, but I think the results would probably be visually indistinguishable and it's worth trying.

**Comment by jchnkl** [comment]
@paolodepetrillo Thank you very much, that's very helpful! The first part sounds pretty straight forward. I'll look into reading the opcodes and then putting them into the gain maps structure. From there on we can experiment to figure out what's the best place for the correction.

**Comment by Mr-Groch** [comment]
Hi! I've also suffering from this issue as Mini 3 owner. Are there any plans to put this on roadmap?

**Comment by Mr-Groch** [comment]
I don't see Mini 3 support added in master branch

**Comment by Mr-Groch** [comment]
Still no Mini 3 DNGs support

**Comment by hennikul** [comment]
I have a DJI Mini Pro 4 Drone, and using Darktable to edit DNG files from that Drone is quite horrible. I have tried to point Claude Sonnet to this issue and asking it to implement support for Opcodelist 3, and that resulted in this change: 

https://github.com/darktable-org/darktable/pull/20467

I have tried this on my drone DNGs, and they now look correct both regarding color and vignetting, so I can finally edit them in Darktable without cumbersome workarounds like layer masks for color balancing etc. 

Now I do admit that this is entirely AI generated, as I have not done C coding for years, and have no experience with Opcodelists, CL-kernels or anything like that. I'm more into Golang, Ruby and Javascript for web systems. But maybe you could look into the code and see if it looks good, or at least could work as a start for implementing this, as Darktable really needs to support this to make sense for editing DNG files from DJI drones (and other DNGs with Opcodelist 3).
